CRM/MembershipExtras/Upgrader.php

```diff
@@ -220,6 +221,162 @@ private function createPeriodLinkCustomFields() {
+      return [];
+    }
+
+    $result = civicrm_api3('Contribution', 'getsingle', [
+      'options' => ['sort' => 'id DESC'],
+      'contribution_recur_id' => $paymentPlanId,
+    ]);
+  }
+
+  /**
+   * @param string $contributionId
+   * 
+   * @return array
+   */
+  private function getLineItemsForContribution($contributionId) {
+    $result = civicrm_api3('LineItem', 'get', [
+      'sequential' => 1,
+      'options' => ['limit' => 0],
+      'contribution_id' => $contributionId,
+    ]);
+
+    $lineItems = [];
+    if ($result['count'] > 0) {
+      $lineItems = $result['values'];
+    }
+
+    return $lineItems;
+  }
+
+  /**
+   * 
+   */
+  private function copyLastInstalmentLineItemsToRecurContrib($paymentPlan, $lineItems) {
+    foreach ($lineItems as $lineItem) {
+      unset($lineItem['id']);
+      unset($lineItem['contribution_id']);
+
+      $params = array_merge($lineItem, [
+        'api.ContributionRecurLineItem.create' => [
+          'contribution_recur_id' => $paymentPlan['id'],
+          'line_item_id' => '$value.id',
+          'start_date' => $paymentPlan['start_date'],
+          'end_date' => $paymentPlan['end_date'],
+          'auto_renew' => $paymentPlan['auto_renew'],
+          'is_removed' => 0,
+        ],
+      ]);
+      civicrm_api3('LineItem', 'create', $params);
+    }
+  }
+
+  private function getCustomFieldId($customGroupName, $customFieldName) {
+    return civicrm_api3('CustomField', 'getvalue', [
+      'return' => 'id',
+      'custom_group_id' => $customGroupName,
+      'name' => $customFieldName,
+    ]);
+  }
+
+  private function createCustomValueForPaymentPlan($paymentPlanId) {
+    $nextPeriodCustomFieldId = $this->getCustomFieldId('related_payment_plan_periods', 'next_period');
+    $prevPeriodCustomFieldId = $this->getCustomFieldId('related_payment_plan_periods', 'previous_period');
+
+    civicrm_api3('ContributionRecur', 'create', [
+      [
+        'id' => $paymentPlanId,
+        'custom_' . $nextPeriodCustomFieldId => 0,
+      ], [
+        'id' => $paymentPlanId,
+        'custom_' . $prevPeriodCustomFieldId => 0,
+      ],
+    ]);
+  }
+
+  /**
+   * Finds all payment plans and populate the offline recurring contribution
+   * line item for the payment plans using an offline payment processor
+   */
+  private function updatePaymentPlans() {
+    $recurContributions = $this->getAllPaymentPlans();
+
+    foreach ($recurringContributions as $paymentPlan) {
+      if ($this->isManualPaymentPlan($paymentPlan['payment_processor_id'])) {
+        $lastInstalment = $this->getLastInstalmentForPaymentPlan($paymentPlan['id']);
+        $lineItems = $this->getLineItemsForContribution($lastInstalment['id']);
+
+        $this->copyLastInstalmentLineItemsToRecurContrib($paymentPlan, $lineItems);
+      }
+
+      $isMatch = (
+        $this->isManualPaymentPlan($paymentPlan['payment_processor_id']) &&
+        !empty($this->getLastInstalmentForPaymentPlan($paymentPlan['id'])) &&
```

Comment:
You're querying for the same contribution twice. Might be better to use the result in the variable and use it here, as `!empty($lastInstalment)`. Every call to the API to the DB is expensive, and since this needs to run for all existing payment plans, there could be a lot of them, so we need to try to be as efficient as possible.